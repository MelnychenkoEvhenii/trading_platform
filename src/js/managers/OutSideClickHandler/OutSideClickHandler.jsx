import React, { useEffect, useRef } from 'react'
import PropTypes from 'prop-types'

const propTypes = {
  handler: PropTypes.func,
  toCapture: PropTypes.bool,
}

const defaultProps = {
  toCapture: false,
}

export const OutSideClickHandler = props => {
  const { handler, children, className, toCapture } = props

  const wrapperRef = useRef(null)

  const useOnClickOutside = (ref, handler) => {
    useEffect(() => {
      const listener = event => {
        if (!ref.current || ref.current.contains(event.target)) {
          return
        }

        handler(event)
      }
      document.addEventListener('mousedown', listener, toCapture)
      document.addEventListener('touchstart', listener, toCapture)
      return () => {
        document.removeEventListener('mousedown', listener, toCapture)
        document.removeEventListener('touchstart', listener, toCapture)
      }
    }, [ref, handler])
  }

  useOnClickOutside(wrapperRef, handler)

  return (
    <div className={className} ref={wrapperRef}>
      {children}
    </div>
  )
}

OutSideClickHandler.propTypes = propTypes
OutSideClickHandler.defaultProps = defaultProps
