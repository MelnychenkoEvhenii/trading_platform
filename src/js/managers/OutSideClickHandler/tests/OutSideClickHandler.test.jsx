import React from 'react';
import { render, cleanup } from 'react-testing-library';
import 'jest-styled-components';

import { OutSideClickHandler } from '../OutSideClickHandler';

const setup = component => {
  const { container } = render(component);
  return container;
};

afterEach(cleanup);

describe('components', () => {
  describe('<OutSideClickHandler />', () => {
    it('should render and match the snapshot', () => {
      expect(setup(<OutSideClickHandler />)).toMatchSnapshot();
    });
  });
});
