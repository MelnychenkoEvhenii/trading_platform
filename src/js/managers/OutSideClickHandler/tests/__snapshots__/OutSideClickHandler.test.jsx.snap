// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components <OutSideClickHandler /> should render and match the snapshot 1`] = `
<div>
  <div />
</div>
`;
